test/controllers/tasks_controller_test.rb

```diff
@@ -105,24 +110,84 @@
     # Note:  If there was a way to fail to save the changes to a task, that would be a great
     #        thing to test.
     it "can update an existing task" do
-      skip
-      # Your code here
+      # Arrange
+      test_id = Task.last.id
+      task_hash = {
+        task: {
+          name: "updated task",
+          description: "updated task description",
+          priority_level: "update priority level",
+        },
+      }
+
+      expect {
+        patch task_path(test_id), params: task_hash
+      }.wont_change "Task.count"
+
+      updated_task = Task.find_by(name: task_hash[:task][:name])
+
+      expect(updated_task.description).must_equal task_hash[:task][:description]
+
+      must_respond_with :redirect
+      must_redirect_to task_path(test_id)
     end
 
     it "will redirect to the root page if given an invalid id" do
-      skip
-      # Your code here
+
+      # Act
+      patch task_path(-1)
+
+      # Assert
+      must_respond_with :redirect
+      expect(flash[:error]).must_equal "Could not find task with id: -1"
+
+      must_redirect_to root_path
     end
   end
 
   # Complete these tests for Wave 4
   describe "destroy" do
-    # Your tests go here
+    it "removes a task from the database" do
+      # Act-Assert
+      task = Task.create!(name: "Sort socks", description: "Find all the lonely socks a pair.", priority_level: "Low")
 
+      expect {
+        delete task_path(task)
+      }.must_change "Task.count", -1
+
+      must_respond_with :redirect
+      must_redirect_to root_path
+
+      after_task = Task.find_by(id: task.id)
+      expect(after_task).must_be_nil
+    end
   end
 
-  # Complete for Wave 4
   describe "toggle_complete" do
-    # Your tests go here
+    it "can mark an incomplete task complete without changing anything else" do
+      # Arrange
```

Comment:
Two missing cases here:
- Does it go the other direction (complete -> incomplete)?
- What happens if the task doesn't exist?